addons/mail/models/mail_message.py

```diff
@@ -1171,10 +1171,17 @@ def _notify_message_notification_update(self):
                     continue
                 else:
                     messages |= message
+        messages_per_partner = defaultdict(lambda: self.env['mail.message'])
+        for message in messages:
+            partner = self.env.user.partner_id
+            author = message.author_id
+            messages_per_partner[author] |= message
+            if author != partner:
+                messages_per_partner[partner] |= message
         updates = [[
-            (self._cr.dbname, 'res.partner', author.id),
-            {'type': 'message_notification_update', 'elements': self.env['mail.message'].concat(*author_messages)._message_notification_format()}
-        ] for author, author_messages in groupby(messages.sorted('author_id'), itemgetter('author_id'))]
+            (self._cr.dbname, 'res.partner', partner.id),
+            {'type': 'message_notification_update', 'elements': message._message_notification_format()}
+        ] for partner, message in messages_per_partner.items()]
```

Comment:
Would name message "messages", with an 's' as you may have several messages per partner :)